### ISSUE TITLE
Zero length Lex item returned from '<unlexable>'

### ISSUE THREAD

**Issue by halostatue** [sqlfluff/sqlfluff#1535]
SQLFluff version 0.6.8.
Python 3.9.7 (via pipx venv)

`.sqlfluff` config:

```
[sqlfluff]
dialect = postgres
output_line_length = 100

[sqlfluff:rules]
tab_space_size = 2
max_line_length = 100

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L014]  # Unquoted identifiers
allow_scalar = true
extended_capitalisation_policy = consistent

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = lower

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = upper
```

Dialect `postgres`

## Expected Behaviour

Cannot parse

```sql
-- The holding account is created using the config.default_users mix task.
CREATE OR REPLACE FUNCTION hold_account_id()
RETURNS uuid LANGUAGE plpgsql VOLATILE STRICT
AS $$
DECLARE
  hold_account_id users.id%TYPE;

BEGIN
  SELECT id INTO hold_account_id
    FROM users
   WHERE email = 'example@example.com';

  IF hold_account_id IS NULL THEN
    RAISE EXCEPTION 'Cannot find user for example@example.com';
  END IF;

  RETURN hold_account_id;
END;
$$;

\prompt 'Cash Account #: ' account
\prompt 'Region (1 - quebec, 2 - east, 3 - west): ' region_number

SELECT *
  FROM card_matches
 WHERE region::text = CASE :region_number WHEN 1 THEN 'quebec' WHEN 2 THEN 'east' WHEN 3 THEN 'west' ELSE 'unknown' END
   AND customer_number = rpad(lpad((:account)::text, 10, '0'), 20, '0');

-- Information to feed into GraphQL.
SELECT 'reassignCashCard' AS mutation,
       encode_relay_id(hold_account_id(), 'Customer') AS new,
       encode_relay_id(cc.customer_id, 'Customer') AS old,
       encode_relay_id(cc.id, 'CustomerCard') AS card
  FROM customer_cards AS cc
 WHERE cc.region::text = CASE :region_number WHEN 1 THEN 'quebec' WHEN 2 THEN 'east' WHEN 3 THEN 'west' ELSE 'unknown' END
   AND cc.code = rpad(lpad(:'account', 10, '0'), 20, '0')
   AND NOT national;
```

## Observed Behaviour

```
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug.
== [priv/repo/sql/unlink-mms-cash-account.sql] FAIL
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0
All Finished 📜 🎉!
```

```
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug. 
...Failed to Parse...
==== parsing violations ====
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0
```

**Comment by halostatue** [comment]
Note that even after these issues are fixed, the issues regarding psql variables in #1534 will still remain, I think.

**Comment by halostatue** [comment]
I can open a new issue for `CASE` support, but it will likely be much later today. With respect to contributing code, that’s unlikely to happen due to severe time constraints on my part that are likely to last until the new year.

I’m not likely to adopt SQLFluff until it can support the column formatting style (I’m an old stick-in-the-mud who has written SQL statements that way for twenty-odd years, and it feels like the One True Way).

It has been an interesting exploration for a few hours, and I’m happy to provide bug reports as I see them (I have one multi CTE query that keeps getting stuck in a formatting / error loop; the first pass over indents a condition and the second pass under indents the condition; rinse and repeat).

**Comment by tunetheweb** [comment]
Closing as original problem is resolved. Should raise another issue for `CASE` support if required.
